### PR TITLE
Suggest defining variable as mutable on `&mut _` type mismatch in pats

### PR DESCRIPTION
**Pull request by WaffleLapkin**
Suggest writing `mut a` where `&mut a` was written but a non-ref type provided.

Since we still don't have "apply either one of the suggestions but not both" kind of thing, the interaction with the suggestion of removing `&[mut]` or moving it to the type is weird, and idk how to make it better..

r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
CI is broken FYI

File: compiler/rustc_typeck/src/check/pat.rs [hunk 1]
```diff
@@ -663,6 +663,46 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        | PatKind::Range(..) => None,
+                    },
+
+                    // Don't provide suggestions in other cases
+                    _ => None,
+                };
+
+                ident_kind.map(|thing| (
+                    pat.span,
+                    format!("to declare a mutable {thing} use `mut variable_name`"),
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/pat.rs:700]: I prefer `to declare a mutable {thing} use:` -- the `mut variable_name` part is redundant with the suggestion that follows direectly after

File: compiler/rustc_typeck/src/check/pat.rs [hunk 2]
```diff
@@ -663,6 +663,46 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                        | PatKind::Path(..)
+                        | PatKind::Box(..)
+                        | PatKind::Ref(..)
+                        | PatKind::Lit(..)
+                        | PatKind::Range(..) => None,
+                    },
+
+                    // Don't provide suggestions in other cases
+                    _ => None,
+                };
```
Comment (compiler-errors) [inline compiler/rustc_typeck/src/check/pat.rs:696]:
```suggestion
                let ident_kind = match binding_parent {
                    hir::Node::Param(_) => "parameter",
                    hir::Node::Local(_) => "variable",
                    hir::Node::Arm(_) => "binding",

                    // Provide diagnostics only if the parent pattern is struct-like,
                    // i.e. where `mut binding` makes sense
                    hir::Node::Pat(Pat { kind, .. }) => match kind {
                        PatKind::Struct(..)
                        | PatKind::TupleStruct(..)
                        | PatKind::Or(..)
                        | PatKind::Tuple(..)
                        | PatKind::Slice(..) => "binding",

                        PatKind::Wild
                        | PatKind::Binding(..)
                        | PatKind::Path(..)
                        | PatKind::Box(..)
                        | PatKind::Ref(..)
                        | PatKind::Lit(..)
                        | PatKind::Range(..) => break 'block None,
                    },

                    // Don't provide suggestions in other cases
                    _ => break 'block None,
                };
```

**Review by compiler-errors — COMMENTED** [review]
Looks fine after nits

File: compiler/rustc_typeck/src/check/pat.rs (hunk 2 above)
Comment (WaffleLapkin) [reply compiler/rustc_typeck/src/check/pat.rs]: Oh, right. I've rewritten this code a couple of times and the match was added before the block xd

**Comment by compiler-errors** [comment]
Cool, thanks.

@bors r+ rollup

**Comment by camelid** [comment]
> Since we still don't have "apply either one of the suggestions but not both" kind of thing, the interaction with the suggestion of removing `&[mut]` or moving it to the type is weird, and idk how to make it better..

I think that actually has been implemented: [`span_suggestions()`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic_builder/struct.DiagnosticBuilder.html#method.span_suggestions)

**Comment by compiler-errors** [comment]
Does `span_suggestions()` work correctly with `Applicability::MachineApplicable`?

**Comment by camelid** [comment]
> Does `span_suggestions()` work correctly with `Applicability::MachineApplicable`?

In what regard? I think rustfix won't autoapply the suggestions if you use `span_suggestions` (since it doesn't know which one you want).

**Comment by WaffleLapkin** [comment]
@camelid rustfix tries to apply all suggestions:
```text
; cargo +stage1 fix --broken-code
    Checking t v0.1.0 (/home/waffle/projects/repos/t)
warning: error applying suggestions to `src/main.rs`

The full error message was:

> Cannot replace slice of data that was already replaced

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

       Fixed src/main.rs (0 fixes)
    Finished dev [unoptimized + debuginfo] target(s) in 0.15s
```

Also the output of the diagnostic is not perfect either, since suggestions don't have labels and it's hard to distinguish them:

```text
error[E0308]: mismatched types
 --> src/main.rs:1:7
  |
1 | fn _f(&mut _a: i32) {}
  |       ^^^^^^^  --- expected due to this
  |       |
  |       expected `i32`, found `&mut _`
  |
  = note:           expected type `i32`
          found mutable reference `&mut _`
help: consider removing `&mut` from the pattern or using `mut` to declare variable as mutable
  |
1 - fn _f(&mut _a: i32) {}
1 + fn _f(_a: i32) {}
  |
1 | fn _f(mut _a: i32) {}
  |       ~~~
```

I think to fix this problem we need a different format for suggestions, that would allow for more flexibility. For example marking placeholders as such (so the tooling could allow users to fill them) or having "either one of these suggestions, but not all" kind of thing. I was meaning to try implementing something like this for a while, but haven't yet.

**Comment by camelid** [comment]
> or having "either one of these suggestions, but not all" kind of thing

I believe that's what `span_suggestions` is supposed to be, but the implementation is non-optimal. In this case, I think you should be able to fix the crash by making the suggestion **non-**`MachineApplicable` (see rust-lang/rustfix#209). When @compiler-errors asked in https://github.com/rust-lang/rust/pull/98431#issuecomment-1165986536 about `MachineApplicable`, I'd forgotten about this issue with rustfix 😅

**Pull request merged**